### ISSUE TITLE
`import/no-unresolved` rule doesn't work properly with webpack resolver

### ISSUE THREAD

**Issue by FezVrasta** [AtomLinter/linter-eslint#610]
I have reported it to `eslint-plugin-import` but probably it's a bug of this package.

https://github.com/benmosher/eslint-plugin-import/issues/376

Doing `$(npm bin)/eslint . --ignore-path .gitignore` everything works fine.
When, instead, I open Atom, it reports `Unable to resolve path to module 'utils/generateHtmlId'`.

I wonder where the problem is...
I'm using the default settings of the package.

**Comment by Arcanemagus** [comment]
What exactly is "report"ing that message?

**Comment by FezVrasta** [comment]
What do you mean? I've installed `linter-eslint` on Atom and using `react-boilerplate` as base for my project.

It reports the error I've reported in the first message here.

**Comment by Arcanemagus** [comment]
This package and `eslint-plugin-import` have nothing to do with `utils/generateHtmlId`, so I have no idea where you are seeing that message, hence why I asked you what is reporting that message to you.

Are you seeing it as an Atom error? A message from `eslint-plugin-import` about a missing module in your code?

**Comment by Arcanemagus** [comment]
Reading the issue @jonase referenced this from, it sounds like this is potentially an issue with your `eslint` configuration. If you want me to take a look at this please provide a zip of a folder structure that shows this, otherwise so far it sounds like this is an external issue.

**Comment by jonase** [comment]
@Arcanemagus Thanks for looking into this. I've tried to create a minimal project (attached). Let me know if you need anything else.

[issue610.zip](https://github.com/AtomLinter/linter-eslint/files/316740/issue610.zip)

**Comment by FezVrasta** [comment]
I've used https://github.com/mxstbr/react-boilerplate, you can see the problem even with that basic boilerplate.

**Comment by Arcanemagus** [comment]
@jonase I'm seeing the same results on the CLI and within Atom for that...

![image](https://cloud.githubusercontent.com/assets/427137/16095259/bf9b6590-32f8-11e6-9ce8-069bd836296a.png)

**Comment by jonase** [comment]
@Arcanemagus I see the same thing with `eslint-plugin-import@1.9.0`. 
`npm install` installed version `1.8.1` and that's where I see the different behaviour. I'll keep investigating this.. the problem might very well be elsewhere. Thanks again.

**Comment by Arcanemagus** [comment]
Hmmm, installing `eslint-plugin-import@1.8.1` _does_ show the behavior described above, I'll see if I can trace it down a bit further.

**Comment by FezVrasta** [comment]
I rep the problem with both versions, but the maintainer told me that 1.9 is bugged. So I guess the investigation should be focused on 1.8

**Comment by IanVS** [comment]
@FezVrasta what version of Atom are you using?  I just updated to the latest beta version, and am seeing some other problems around eslint-plugin-import.

**Comment by FezVrasta** [comment]
```
$ atom -v && apm -v
Atom    : 1.8.0
Electron: 0.36.8
Chrome  : 47.0.2526.110
Node    : 5.1.1
apm  1.9.2
npm  2.13.3
node 0.10.40
python 2.7.9
git 2.6.4
```

**Comment by FezVrasta** [comment]
Still no solutions?

I can confirm that it's absolutely a bug of this package, because using the command line everything works perfectly.

**Comment by iofjuupasli** [comment]
> I can confirm that it's absolutely a bug of this package

I can confirm it too. It's unlikely bug on my side.

Running eslint from terminal doesn't report any error, but in atom it shows `import/no-unresolved` error.

This is how I run eslint (using npm scripts):
```
eslint ./
```

Plugin uses local eslint (same as in terminal)

```
Atom    : 1.8.0
Electron: 0.36.8
Chrome  : 47.0.2526.110
Node    : 5.1.1
apm  1.9.2
npm  2.13.3
node 0.10.40
python 2.7.10
git 2.7.4
```

**Comment by stoikerty** [comment]
I'm having the same issue in Atom, but I'm also getting the issue on the command-line when I run `eslint src`. When compiling it through webpack (which uses `eslint-loader`) everything works fine though.

relevant part of my `.eslintrc` config
```js
  "settings" : {
    "import/resolver": {
      "webpack": {
        "config": "toolkit/webpack/config.js"
      }
    }
  },
```

relevant part of my webpack config
```js
  resolve: {
    extensions: ['', '.js', '.jsx'],
    root: path.resolve(__dirname, root),
    alias: {
      src: 'src',
      client: 'src/client',
      server: 'src/client',
    },
    modulesDirectories: [
      clientRoot,
      'node_modules',
    ],
  },
```

**Comment by stoikerty** [comment]
I think it has to do with using a custom `config`-path in `.eslintrc`?

**Comment by IanVS** [comment]
@stoikerty I think you're on to something, in that it might have something to do with Webpack.  The code that @jonase posted is also using `"import/resolver": "webpack"`, but no custom `config` path.  @iofjuupasli and @FezVrasta can you confirm whether you're using `"import/resolver": "webpack"`?

**Comment by stoikerty** [comment]
As a temporary fix, I've disabled the rule for imports starting in `src/` using:

```js
  "rules": {
    "import/no-unresolved": [
      2,
      {
        "ignore": [ 'src/' ]
      }
    ]
  },
```

**Comment by FezVrasta** [comment]
Yes I'm using the webpack resolver

**Comment by nikolas6** [comment]
I am having the same problem with linter-eslint (Atom1.9.4) shows `"import/no-unresolved"` when importing from local module. No problem importing modules from `node_modules` folder tho.

SublimeText or linting directly from the command line with `"eslint ."` works fine.

I only see this problem on my Windows7 machine btw. Everything works fine on Mac.

**Comment by Gyllsdorff** [comment]
I had the same problem but I was able to resolve it by removing all relative paths from the resolve section of my `webpack.config.js`. 

I think `linter-eslint` on windows runs `eslint` using a different working directory so every path using non absolute paths will be incorrect. If I use eslint from a cmd window using `node_modules\.bin\eslint src/app` it works even if I use relative paths.

I changed `root: path.resolve('./src/app')` to `root: path.resolve(__dirname, 'src', 'app')` in my `webpack.config.js` and everything started working after I reloaded Atom.

OS: Windows 10.
Atom version: 1.9.5
Linter-eslint: 7.2.4
eslint: 3.2.2
eslint-plugin-import: 1.12.0
eslint-import-resolver-webpack: 0.4.0

**Comment by FezVrasta** [comment]
I'm having the problem on Mac, so I doubt it's a Windows only issue.

**Comment by nikolas6** [comment]
I managed to find the culprit that was causing the problem for me. I have an external json file where I keep list of application settings, which I call in my webpack config file `const settings = require('../../settings.json')`.
Not sure why, but having that cause the `"import/no-unresolved"` error on my Windows7.

I modified it to `require('./../../settings.json')`, and everything is fine now.

**Comment by FezVrasta** [comment]
Seems a pretty horrible workaround, the problem is still to get fixed on this library

**Comment by IanVS** [comment]
@FezVrasta if this is important to you, I'd suggest taking a look at the code and seeing if you can make a fix.  That's the beauty of OSS, after all.  I don't use Webpack, and I don't think that @Arcanemagus does either, so it's a bit more difficult for us to reproduce and make a fix.  It seems that there's an issue with cwd, which this package sets to either an `.eslintignore` file directory, or the directory of the file being processed (https://github.com/AtomLinter/linter-eslint/blob/master/src/worker-helpers.js#L92-L95).  I have no idea how webpack works, but maybe that's causing some issues?

**Comment by FezVrasta** [comment]
I hope I had free time for this! I'll start to work with React and Webpack again soon, if I'll encounter this problem even with the new environment I'll surely dedicate some time to fix it.

**Comment by easingthemes** [comment]
I'm having this issue without Atom, just `eslint` and `webpack` from command line. So maybe it's not Atom related.

```
"eslint": "^3.0.1", 
"webpack": "^1.12.14",
```

**Comment by FezVrasta** [comment]
@easingthemes yours must be a different issue, because me (and others) already tested the CLI tool and it works as expected.

**Comment by Neophy7e** [comment]
@FezVrasta @IanVS  The only fix that I found for now in order to not have the issue is to create an empty .eslintignore file at the root directory of your project

**Comment by iofjuupasli** [comment]
Using `"moduleDirectories"` instead of `"paths"` fixes problem for me

**Comment by IanVS** [comment]
@Neophy7e that's really interesting, and gives a good clue.  When linter-eslint finds an `.eslintignore` file, it sets `cwd` to that location.  

@FezVrasta There's been reports of at least one other person having trouble from the command line.  

@easingthemes, do you have the cli problems when you run linting from somewhere other than your project root?

**Comment by nrdobie** [comment]
@IanVS Adding a blank `.eslintignore` file to my project fixed the issues with alias. But even when I set an absolute path to my config it still failed without the `.eslintignore`, which I find extremely weird.

**Comment by IanVS** [comment]
Thanks @nrdobie.  I've started using webpack recently, so I should be able to dig into this fairly soon.  (It's working fine for me so far, perhaps because I have an `.eslintignore` file.)

**Comment by tiodan81** [comment]
I've been experiencing the same problem. Imports in all .js files that are not in the project root (i.e. everything in `./client`) throw `Resolve error: Module version mismatch. Expected 49, got 48.` in Atom but not when running eslint from the command line. Files in the project root (e.g. the webpack config) report no errors. 

Using Atom 1.12.5
eslint 3.9.1
eslint-import-resolver-webpack 0.6.0
webpack 2.1.0
linter-eslint 8.0.0

**Comment by mariusc23** [comment]
Adding `.eslintignore` to project root fixed it for me, but I had to restart Atom.

**Comment by bobbypriambodo** [comment]
Wow, adding blank `.eslintignore` and reloading Atom (Ctrl+Alt+Cmd+L) really fixes the issue on my Mac. Thanks @Neophy7e!

AFAIK `.eslintignore` is not required to be on every project. Any reason why it is used to determine the `cwd`? Why not `.eslintrc.*`?

And if I remember correctly, this used to be not an issue. I think it was after I changed my webpack config resolve to using `root` instead of `modulesDirectories` (since the recursive nature of `modulesDirectories` introduced a bug on one of my deps) that the errors popped up, and somehow they didn't go away when I tried reverting the config.

**Comment by IanVS** [comment]
>AFAIK `.eslintignore` is not required to be on every project. Any reason why it is used to determine the `cwd`? Why not `.eslintrc.*`

You are absolutely correct that it's not required to use a `.eslintignore` file, but it does tend to be a pretty good indicator of a project root.  Config files (`.eslintrc`) cascade, and therefore are often scattered throughout a project. On the other hand, `.eslintignore` files do not cascade, and are normally placed at a project's root directory.

More importantly, until https://github.com/eslint/eslint/pull/7678 is merged (breaking change, waiting for eslint v4), the patterns in an `.eslintignore` file are always treated relative to `cwd`.  So, we (`linter-eslint`) need to make sure to set `cwd` to the folder containing the ignore file for the patterns to be applied as-expected. 

It seems that the webpack resolver has some dependence on `cwd` being set to the project root.  If there is no `.eslintignore` file, then `linter-eslint` does not have a reliable heuristic for determining project root.  We could possibly consider the folder which is open in the editor to be the project root, which will be true in most cases, but if a user opens just an individual file that will not work either.

I think some investigation will need to be done into the webpack resolver, what its requirements are for `cwd`, and whether that behavior can be changed.

**Comment by tero** [comment]
Not sure if same issue but I found that following lines in .eslintrc.json stops linter-eslint from linting:
```
    "settings": {
      "import/resolver": {
        "webpack": {
          "config": "webpack.config.babel.js"
        }
      }
    }
```

When running eslint from command line it reports few issues but Atom doesn't report anything anymore. If I remove that block from config, Atom is line with eslint cli, but both can't find our local modules.

**Comment by rpowelll** [comment]
so here's something interesting, when I run `./node_modules/.bin/eslint src/` from my project's root directory, I don't see any errors related to unresolved imports, but when I `cd` into the `src/` directory and run `../node_modules/.bin/eslint .`, I see a bunch of errors. So the root of this problem might have something to do with _linter-eslint_ not running `eslint` from the project's root directory?

**Comment by webdif** [comment]
I'm not using webpack, but Brunch. So I use `eslint-import-resolver-node`, and I have the same linter problem: without the `.eslintignore` file, I have errors. When a create a empty file, the error is gone.

**Comment by shidaping** [comment]
just config eslint-import-plugin like below instead of webpack:
`{
  "settings": {
    "import/resolver": {
      "node": {
        "paths": ["src"]
      }
    }
  }
}`
this works for me

**Comment by nmuldavin** [comment]
I solved this problem by specifying the full absolute path to the directory in my `.eslintrc.js`. @shidaping's solution wasn't enough

```
const path = require('path');

module.exports = {
  ...(Other eslint config),
  settings: {
      'import/resolver': {
        node: {
          paths: [path.resolve(__dirname, 'src')],
        },
      },
  },
}
```

**Comment by fernandofleury** [comment]
https://github.com/benmosher/eslint-plugin-import/issues/799 Maybe this is related?

I've had to stop using the webpack resolver and use node as the resolver.

**Comment by landonalder** [comment]
The only solution I found was to use `eslint-import-resolver-babel-module` instead of the webpack resolver. Since I have my aliases inside my .babelrc, it worked like a charm.

**Comment by Arcanemagus** [comment]
I'm pushing this back on `eslint-import-resolver-webpack` being broken (as the comments in https://github.com/benmosher/eslint-plugin-import/issues/376 seem to indicate is still the case as well).

On the plus side, on retesting this issue with current versions of all packages we are now seeing the exact same results that ESLint itself gives you, so I'm marking this issue as closed.

Atom:
![image](https://user-images.githubusercontent.com/427137/30518407-6007747c-9b32-11e7-9c24-0ab6a1102ada.png)

ESLint CLI:
![image](https://user-images.githubusercontent.com/427137/30518412-6f9706f0-9b32-11e7-9f6c-78a339a48c57.png)

**Comment by bombillazo** [comment]
> 
> 
> As a temporary fix, I've disabled the rule for imports starting in `src/` using:
> 
> ```js
>   "rules": {
>     "import/no-unresolved": [
>       "error",
>       {
>         "ignore": [ 'src/' ]
>       }
>     ]
>   },
> ```

This did it for my babel config for React Native config files:
